cng/rsa.go

```diff
@@ -170,30 +168,12 @@ func encodeRSAKey(N, E, D, P, Q, Dp, Dq, Qinv BigInt) ([]byte, error) {
 	}
 	copy(blob, (*(*[sizeOfRSABlobHeader]byte)(unsafe.Pointer(&hdr)))[:])
 	data := blob[sizeOfRSABlobHeader:]
-	var err error
-	encode := func(b BigInt, size uint32) {
-		if err != nil {
-			return
-		}
-		// b might be shorter than size if the original big number contained leading zeros.
-		leadingZeros := int(size) - len(b)
-		if leadingZeros < 0 {
-			err = errors.New("crypto/rsa: invalid parameters")
-			return
-		}
-		copy(data[leadingZeros:], b)
-		data = data[size:]
-	}
-	encode(E, hdr.PublicExpSize)
-	encode(N, hdr.ModulusSize)
-	if D != nil {
-		encode(P, hdr.Prime1Size)
-		encode(Q, hdr.Prime2Size)
-		encode(Dp, hdr.Prime1Size)
-		encode(Dq, hdr.Prime2Size)
-		encode(Qinv, hdr.Prime1Size)
-		encode(D, hdr.ModulusSize)
-	}
+	err := encodeBigInt(data, []sizedBigInt{
+		{E, hdr.PublicExpSize}, {N, hdr.ModulusSize},
+		{P, hdr.Prime1Size}, {Q, hdr.Prime2Size},
+		{Dp, hdr.Prime1Size}, {Dq, hdr.Prime2Size},
+		{Qinv, hdr.Prime1Size}, {D, hdr.ModulusSize},
+	})
```

Comment:
Hmm, I assume that if `D == nil`, then `P` etc. must be `nil`? I like the explicitness that was here before with `if D != nil`, considering we also used `if D == nil` to allocate `blob` earlier. It looks like if `P != nil` but `D == nil`, we might hit some weird behavior when it tries to encode `P`, `Q`, etc. (although maybe just a panic). I don't anywhere `P`, `Q`, ... are all checked for `nil`ness earlier in this package.

(Putting the `nil` check in `encodeBigInt` seems to me to make `encodeRSAKey` and `encodeECCKey` a little less clear about which `nil`s are expected. I think pulling the check out and actually having `encodeBigInt` panic on nil would be a bit more intuitive than stopping on first nil. But: `encodeBigInt` is documented well and isn't exported, so if it helps perf or reusability, I'm not opposed to leaving the `nil` check in there.)

Reply:
Good catch with the missing `D != nil` check in the new approach. CNG does not support empty P, Q, Dp, Dq, or Qinv when D is set, so I'll be even more explicit and check for this error condition before trying to encode the key.